ffi/diplomat/src/collator.rs

```diff
@@ -0,0 +1,238 @@
+        pub case_level: ICU4XCollatorCaseLevel,
+        pub numeric: ICU4XCollatorNumeric,
+        pub backward_second_level: ICU4XCollatorBackwardSecondLevel,
+    }
+
+    #[non_exhaustive]
+    #[diplomat::rust_link(icu_collator::Strength, Enum)]
+    pub enum ICU4XCollatorStrength {
+        Auto = 0,
+        Primary = 1,
+        Secondary = 2,
+        Tertiary = 3,
+        Quaternary = 4,
+        Identical = 5,
+    }
+
+    #[non_exhaustive]
+    #[diplomat::rust_link(icu_collator::AlternateHandling, Enum)]
+    pub enum ICU4XCollatorAlternateHandling {
+        Auto = 0,
+        NonIgnorable = 1,
+        Shifted = 2,
+    }
+
+    #[non_exhaustive]
+    #[diplomat::rust_link(icu_collator::CaseFirst, Enum)]
+    pub enum ICU4XCollatorCaseFirst {
+        Auto = 0,
+        Off = 1,
+        LowerFirst = 2,
+        UpperFirst = 3,
+    }
+
+    #[non_exhaustive]
+    #[diplomat::rust_link(icu_collator::MaxVariable, Enum)]
+    pub enum ICU4XCollatorMaxVariable {
+        Auto = 0,
+        Space = 1,
+        Punctuation = 2,
+        Symbol = 3,
+        Currency = 4,
+    }
+
+    #[non_exhaustive]
+    #[diplomat::rust_link(icu_collator::CaseLevel, Enum)]
+    pub enum ICU4XCollatorCaseLevel {
+        Auto = 0,
+        Off = 1,
+        On = 2,
+    }
+
+    #[non_exhaustive]
+    #[diplomat::rust_link(icu_collator::Numeric, Enum)]
+    pub enum ICU4XCollatorNumeric {
+        Auto = 0,
+        Off = 1,
+        On = 2,
+    }
+
+    #[non_exhaustive]
+    #[diplomat::rust_link(icu_collator::BackwardSecondLevel, Enum)]
+    pub enum ICU4XCollatorBackwardSecondLevel {
+        Auto = 0,
+        Off = 1,
+        On = 2,
+    }
+
+    impl ICU4XCollator {
+        /// Construct a new Collator instance.
+        #[diplomat::rust_link(icu::collator::Collator::try_new_unstable, FnInStruct)]
+        pub fn try_new(
+            provider: &ICU4XDataProvider,
+            locale: &ICU4XLocale,
+            options: ICU4XCollatorOptions,
+        ) -> DiplomatResult<Box<ICU4XCollator>, ICU4XError> {
+            use icu_provider::serde::AsDeserializingBufferProvider;
+            let provider = provider.0.as_deserializing();
+            let locale = locale.to_datalocale();
+            let options = CollatorOptions::from(options);
+
+            Collator::try_new_unstable(&provider, &locale, options)
+                .map(|o| Box::new(ICU4XCollator(o)))
+                .map_err(Into::into)
+                .into()
+        }
+
+        /// Compare guaranteed well-formed UTF-8 strings.
+        #[diplomat::rust_link(icu::collator::Collator::compare, FnInStruct)]
+        pub fn compare(&self, left: &str, right: &str) -> ICU4XOrdering {
+            self.0.compare(left, right).into()
+        }
+
+        /// Compare potentially ill-formed UTF-8 strings.
+        #[diplomat::rust_link(icu::collator::Collator::compare_utf8, FnInStruct)]
+        pub fn compare_utf8(&self, left: &[u8], right: &[u8]) -> ICU4XOrdering {
```

Comment:
Per previous discussion, could you, please make this `left: &str` and `right: &str` and immediately call `as_bytes()` on them, so that both `compare` and `compare_utf8` take `std::string_view` in C++?

Also, would it be bad for code generation for non-C++ languages if `compare` was named `compare_unsafe` and `compare_utf8` was named `compare` (to guide C++ callers who aren't thinking carefully about UTF-8 well-formedness to the UBless option)?

Reply:
Right now the only non-C++ language we have is JS, where we force-convert to UTF8, so calling it unsafe is a bit weird. I think the current model is fine, actually? That previous discussion about `&str`-that-is-bytes was more for the case where we _only_ take in `&str`, here we take in both anyway.

For 1.0 all of the FFI APIs will be suboptimal, I'm hoping to be able to massively improve Diplomat over the next year so that we can have a breaking 2.0 of FFI that is much much better.

Reply:
I think it still makes sense to take `&str` here and immediately call `.as_bytes()` to get the API that makes sense for C++.

Reply:
To elaborate on why I think it makes sense to use `&str` with immediate `.as_bytes()` for `compare_utf8`:

In C++, in the kind of code that uses `std::string_view` (and types that autoconvert to it), the use of `std::string_view` doesn't really signify an UB-if-wrong-level commitment to well-formed UTF-8. That is, `std::string_view` in C++ in contrast to `std::span<uint8_t>` doesn't carry a type-system-level commitment of UTF-8 well-formedness.

Furthermore, it's never semantically wrong to call `compare_utf8` instead of `compare`: Doing so with well-formed inputs is a perf pessimization (I should measure how large!), but that's it. Therefore, we should have documentation and ergonomics that guide C++ caller who haven't thought long and hard about ensuring well-formedness to use `compare_utf8` instead of `compare` without much thinking. This doesn't work if `compare_utf8` has an unidiomatic argument type while `compare` has the idiomatic type (but also has UB).

(I don't like relying on a 2.0 break for something like this when there's such an easy fix available right now (`&str` and immediate `.as_bytes()`).)

Reply:
Yes, but that argument is one-sided from the C++ POV: From the JS POV it's a differently shaped mess. I don't consider that an "easy fix" because there's a tricky tradeoff.

I think in that case we should just _not have_ `compare_utf8` and have `compare` do the validating string thing (and have JS eat the performance impact)

Reply:
I think exposing a function that routes to the function that in Rust is called `compare_utf8` without exposing a function that routes to what in Rust is called `compare` is a reasonable option for now, but then that should still use the `&str` with immediate `.as_bytes()` hack to get the idiomatic signature in C++ and, AFAICT, to get the string conversion behavior in JS.

Going forward, I think it would make sense for Diplomat to have per-target-language directives that would allow exposing only Rust `compare_utf16` to JS, Java, and other languages whose strings are potentially-ill-formed UTF-16, allow exposing only Rust `compare` to Swift (I'm assuming here that Swift guarantees UTF-8 well-formedness), allow exposing only Rust `compare_utf8` to Go, etc. I think it would be appropriate to expose all three to C++ with docs and ergonomics that push the one that routes to Rust `compare_utf8` as what callers should use when in doubt.

Reply:
Okay, that works for now. I'm also open to exposing `compare()` as well but we can start with the simpler thing and add stuff over time.

> Going forward, I think it would make sense for Diplomat to have per-target-language directives that would allow exposing only Rust `compare_utf16` to JS, Java, and other languages whose strings are potentially-ill-formed UTF-16, allow exposing only Rust `compare` to Swift (I'm assuming here that Swift guarantees UTF-8 well-formedness), allow exposing only Rust `compare_utf8` to Go, etc. I think it would be appropriate to expose all three to C++ with docs and ergonomics that push the one that routes to Rust `compare_utf8` as what callers should use when in doubt.

Yes, that's already the plan.

To spell it out, the long term plan is:

 - Diplomat has per-backend rename/disabling (https://github.com/rust-diplomat/diplomat/issues/233, https://github.com/rust-diplomat/diplomat/issues/234) so we can shape the API in a per-backend way
 - Diplomat gets an UnvalidatedStr input type, which in JS is identical to `&str` but in C++ is a `string_view` that has no validation guarantees
 - Diplomat's `&str` type is by-default validated in C++, though we have a global config option to disable that. You can "opt in" to that behavior again by using UnvalidatedStr.

These can be achieved in a non-breaking way if done right, though we need to make sure our choices with the API now are compatible with this. If we don't add a `compare()` now it gives us some breathing room.

Reply:
Okay, so I think we should land this as is and then discuss the details of what to do for 1.0 when I fix https://github.com/unicode-org/icu4x/issues/2520 (which I plan to do after this lands) just so that Elango's work isn't blocked.

I _think_ we already have consensus over the plan but I don't want to drag this PR further